### ISSUE TITLE
gpu-inference: vLLM v0.19 deployment with DRA GPU allocation and Multi-LoRA

### ISSUE THREAD

**Issue by 100rd** [100rd/platform-design#78]
## Summary

Deploy vLLM v0.19.0 inference engine on the gpu-inference cluster with DRA-based GPU allocation, PagedAttention, FlashAttention 4, and Multi-LoRA adapter support.

## Context

vLLM is the inference engine for serving LLMs. Key advantages:
- **PagedAttention**: Efficient VRAM management, near-zero waste on KV cache
- **FlashAttention 4**: New in v0.19, next-gen attention performance
- **Multi-LoRA**: Serve multiple fine-tuned models from a single base model instance → dramatically fewer pods
- **Continuous batching**: Dynamic request batching for optimal throughput

### vLLM Deployment
- [ ] Kubernetes manifests: `apps/gpu-inference/vllm/`
- [ ] Deployment with DRA ResourceClaim:
  ```yaml
  apiVersion: apps/v1
  kind: Deployment
  metadata:
    name: vllm-inference
    namespace: gpu-inference
  spec:
    replicas: 3
    selector:
      matchLabels:
        app: vllm-inference
    template:
      metadata:
        labels:
          app: vllm-inference
      spec:
        schedulerName: volcano
        priorityClassName: gpu-inference-standard
        containers:
        - name: vllm
          image: vllm/vllm-openai:v0.19.0
          ports:
          - containerPort: 8000
            name: http
          - containerPort: 9090
            name: metrics
          args:
          - "--model=/models/meta-llama/Llama-3.1-70B"
          - "--tensor-parallel-size=4"
          - "--max-model-len=32768"
          - "--enable-lora"
          - "--max-loras=8"
          - "--gpu-memory-utilization=0.95"
          - "--enable-chunked-prefill"
          - "--max-num-batched-tokens=65536"
          resources:
            claims:
            - name: gpus
            requests:
              cpu: "16"
              memory: "64Gi"
              hugepages-1Gi: "32Gi"
          volumeMounts:
          - name: model-cache
            mountPath: /models
          - name: hugepages
            mountPath: /dev/hugepages
        resourceClaims:
        - name: gpus
          resourceClaimTemplateName: vllm-gpu-claim
        volumes:
        - name: model-cache
          persistentVolumeClaim:
            claimName: model-cache-pvc
        - name: hugepages
          emptyDir:
            medium: HugePages-1Gi
  ```

### ResourceClaimTemplate for vLLM
- [ ] 4-GPU claim for 70B model (tensor parallel):
  ```yaml
  apiVersion: resource.k8s.io/v1
  kind: ResourceClaimTemplate
  metadata:
    name: vllm-gpu-claim
  spec:
    spec:
      devices:
        requests:
        - name: gpus
          deviceClassName: nvidia-h100-sxm5
          count: 4
  ```

### Multi-LoRA Configuration
- [ ] Base model loaded once, LoRA adapters hot-loaded
- [ ] LoRA adapter storage: S3 bucket with EFS cache
- [ ] API endpoint: `/v1/chat/completions` with `model` parameter routing to LoRA
- [ ] Max 8 concurrent LoRA adapters per instance

### Model Storage
- [ ] PVC for model weights (EFS or FSx for Lustre)
- [ ] Model download init container or pre-cached AMI
- [ ] HuggingFace model hub integration (token via External Secrets)

### Service & Networking
- [ ] ClusterIP Service for internal inference
- [ ] Optional: NLB for external API access
- [ ] gRPC support for high-performance clients
- [ ] Health check: `/health` endpoint

### Prometheus Metrics
- [ ] vLLM exposes metrics on `:9090/metrics`:
  - `vllm:avg_generation_throughput` (tokens/sec)
  - `vllm:num_requests_waiting` (queue depth)
  - `vllm:num_requests_running`
  - `vllm:gpu_cache_usage_perc`
  - `vllm:avg_prompt_throughput`
- [ ] VMServiceScrape for VictoriaMetrics

### Validation
- [ ] vLLM serves completions on H100 GPUs via DRA
- [ ] Multi-LoRA: switch between adapters without restart
- [ ] HugePages utilized (verify via `/proc/meminfo` in pod)
- [ ] Benchmark: throughput > 1000 tokens/sec per 4-GPU instance

## Versions

| Component | Version |
|-----------|---------|
| vLLM | v0.19.0 |
| PyTorch | 2.10.0 |
| FlashAttention | 4.x |
| CUDA | 12.8+ |

## Environment

- **Prod only**

## Dependencies

- DeviceClass definitions (#68) — GPU allocation
- Volcano scheduler (#73) — scheduling
- VictoriaMetrics (#75) — metrics collection
- Node tuning (#66) — HugePages available

## References

- [vLLM docs](https://docs.vllm.ai/en/latest/)
- [vLLM v0.19 release](https://github.com/vllm-project/vllm/releases)
- [vLLM Multi-LoRA](https://docs.vllm.ai/en/latest/features/lora.html)
- [vLLM production deployment](https://www.spheron.network/blog/vllm-production-deployment-2026/)
